### DISCUSSION TITLE
Multi-dimensional arrays on the Pico

### DISCUSSION THREAD

**Discussion by KlaireLarnia** [micropython/micropython#13033] [RP2040 / Pico] [answered]
Hi.

I've tried googling this and I am clearly not grasping how to do this in Micro Python on the Pico.

I need to use an array to hold data read from a text file. In good old BASIC I would have set up a simple multi-dimensional array along the lines of:

DIM X(30,3)

Where by I had 30 sections of 3 values = 90 in total. So I could call X(1,1) to get the first number 1, "X(1,2)" to get next number and X(1,3) to get the last for that row.

All the numbers I need to store are integers and range from 0 to 999. 

How would I set this up in Micro Python as far as I can see there is no DIM command and the closest I can get is a list which will store the numbers sequentially but then I need to do basic maths to work out the position I need to read from and to.

I believe there is an array command but that again doesn't, from what I can tell, allow me to create a dimensional array

**Comment by rkompass** [comment]
Seems you need [ulab](https://micropython-ulab.readthedocs.io/en/latest/). Its like micropythons version of numpy.
You can do something like
```py
a = array([[1, 0, 0],
           [0, 2, 0],
           [0, 0, 3]], dtype=uint8)
```
which is what you wanted?
To use it you need a firmware that has ulab compiled in.

**Reply by KlaireLarnia** [reply]
Thank you both. I am glad I wasn't going mad and missing something. Weird there is no way to do not natively in Python given it's been in BASIC since I was in school over 30-years ago. Clearly things are done very differently these days.

I'll take a look at both options as both seem to offer interesting solutions to what I need to do.

Thank you again :)

**Reply by rkompass** [reply]
> Clearly things are done very differently these days.

Clearly. When I was learning BASIC on a ZX Spectrum and later C (Hisoft C compiler) on that same device the emphasis was that in C most functionality was transferred from the language itself to libraries. Even the printing function suddenly was no longer part of the language, but you had to include a header file for printf().
That allowed the language to remain slim. A concept that survived.
Python tries the same, although it is much larger.

Turns out most, if not all, data analysts now use Numpy or libraries that build on Numpy.
Numpy in turn is a re-implementation of MATLABs array functionality, which was considered to be superior then, for Python.
So you have the best of two worlds: Python - which is a nice language to program in, and MATLABs arrays, without the MATLAB language itself, which is not that comfortable and clear.

**Reply by 2dof** [reply]
'''Numpy in turn is a re-implementation of MATLABs array functionality, which was considered to be superior then, for Python'' 
Both Matlab and Numpy use old well proven LAPACK/BLAS  Libs. The difference is how they are called: Matlab is "build on" matrix operation ( all numbers are seen as matrices also a memory storage optimized). Numpy just  call indirectly LAPACK/BLAS so it is slower,  but if You call Lapack/blass  directly then perfomance is similar for both ( and much faster than Numpy way)
then

**Comment by andrewleech** [comment, marked as answer]
It's true there's no built-in python method for multi-dimensional arrays (on desktop or micropython)

For basic needs like this I'd generally just use a "list of lists", this looks like a good example: https://snakify.org/en/lessons/two_dimensional_lists_arrays/

If you need to do more serious processing, the ulab project might be what you need, they provide firmware builds with ulab included: https://github.com/v923z/micropython-ulab

**Reply by KlaireLarnia** [reply]
Thank you both. I am glad I wasn't going mad and missing something. Weird there is no way to do not natively in Python given it's been in BASIC since I was in school over 30-years ago. Clearly things are done very differently these days.

I'll take a look at both options as both seem to offer interesting solutions to what I need to do.

Thank you again :)

**Reply by peterhinch** [reply]
It is possible to write a class that enables multi-dimensional array syntax. See [this doc](https://github.com/peterhinch/micropython-samples/blob/master/parse2d/README.md). This is slightly more involved than you might expect because of Python's flexible slice notation. With the example two dimensional module you can write
```py
demo = MyClass(10, 10)  # Create an object that can be viewed as 10 rows * 10 cols
demo[8, 8] = 42  # Populate a single cell
demo[0:, 0] = iter((66, 77, 88))  # Populate a column
demo[2:5, 2:5] = iter(range(50, 60))  # Populate a rectangular region.
print(sum(demo[0, 0:]))  # Sum a row
for row, _ in enumerate(demo[0:, 0]):
    print(*demo[row, 0:])  # Size- and shape-agnostic print
```

**Comment by KlaireLarnia** [comment]
Thanks. I went with the "list of lists" idea as that was easy for me to understand and worked perfectly for what I needed.

**Reply by peterhinch** [reply]
Fair enough, that is the standard approach. My aim was to illustrate the flexibility of the Python language.

**Comment by KlaireLarnia** [comment]
I'm already planning the next stage of my project which will move it from a pi pico to my pi4, so I m sure your  comments will help then as I will need to learn new things like classes and better coding principles / disciplines.       What I have now is a proof of concept - evidence I can do what I aimed.

The next evolution which will be next year I hope will be another level above this so I am grateful for all the help as it shows me avenues to look at for that.
